base/vistat/vistat.man
.ds Vn 1.12
.TH vistat 1Vi "24 April 1993" "Vista Version \*(Vn"
.SH NAME
vistat \- report image statistics
.SH SYNOPSIS
\fBvistat\fR [\fB-\fIoption\fR ...] [\fIinfile\fP ...] [\fIoutfile\fP]
.SH DESCRIPTION
\fBvistat\fP reports the minimum, maximum, mean, and standard deviation of
the pixel values in its input images.
.PP
The measurements are reported in the form of a Vista data file containing
one \fBstatistics\fP attribute per image measured. Each \fBstatistics\fP
attribute, in turn, contains \fBmin\fP, \fBmax\fP, \fBmean\fP, and 
\fBstd_dev\fP attributes reporting individual measurements. It also
contains a \fBname\fP attribute if the corresponding image has one.
Because these attribute have no binary data associated with them the
program's output can be viewed directly at a terminal.
.SH "COMMAND LINE OPTIONS"
\fBvistat\fP accepts the following options:
.IP \fB-help\fP 15n
Prints a message describing options.
.IP "\fB-in\fP \fIinfile\fP"
Specifies a Vista data file containing the input images.
.IP "\fB-out\fP \fIoutfile\fP"
Specifies where to write the output statistics as a Vista data file.
.PP
Input and output files can be specified on the command line or allowed to
default to the standard input and output streams.
.SH "SEE ALSO"
.na
.nh
.BR VImageStats (3Vi),
.BR VImage (3Vi),
.BR Vfile (5Vi),
.BR Vista (7Vi)
.hy
.ad
.SH AUTHOR
Art Pope <pope@cs.ubc.ca>


debian/changelog
via (2.0.1svn191-0) stable; urgency=low

  *renamed via-bin to via-utils

 -- Lipsia Group (The official Lipsia developer group) <lipsia@cbs.mpg.de>  Tue, 09 Nov 2010 12:17:00 +0400 

via (2.0.0-0) stable; urgency=low

  *This version comes with the new lipsia-2.0 package
   and comprises some minor bug fixes
  *Fixed type error of VLong (is now long)

 -- Lipsia Group (The official Lipsia developer group) <lipsia@cbs.mpg.de>  Tue, 09 Nov 2010 12:17:00 +0400 

via (1.6.0-2) unstable; urgency=low

  * New internal release

 -- Frank Burkhardt <burk@cbs.mpg.de>  Mon, 20 Jul 2009 11:45:41 +0200

via (1.6.0-1) unstable; urgency=low

  * Initial release (Closes: #521328).
  * New Upstream Version. Renamed 'vcat' to 'vcatobj' to not conflict with
    atfs package in Debian.
  * Update Standards-version to 3.8.2.0 -- no changes necessary.
  * Update watchfile for recent change in upstream naming scheme.
  * Add versioned dependency on debhelper (>= 5) upon recommendation by
    lintian.

 -- Michael Hanke <michael.hanke@gmail.com>  Fri, 19 Jun 2009 13:50:52 +0200

via (1.5.2+svn3529-2) UNRELEASED; urgency=low

  * Add alternative build-dependency on atlas3-base-dev to be able to build
    the package on Debian etch.
  * FTP access to the upstream sources needs passive mode. Fixed debian/watch.

 -- Michael Hanke <michael.hanke@gmail.com>  Tue, 14 Apr 2009 02:46:10 +0200

via (1.5.2+svn3529-1) UNRELEASED; urgency=low

  * New Upstream SVN snapshot that relocates the VIA lib headers to their new
    cannonical location.

 -- Michael Hanke <michael.hanke@gmail.com>  Tue, 31 Mar 2009 22:08:01 +0200

via (1.5.2-1) UNRELEASED; urgency=low

  * Initial Debian packaging. This is a complete rewrite of the original
    Debian packaging effort done by the upstream developers.

 -- Michael Hanke <michael.hanke@gmail.com>  Thu, 26 Mar 2009 19:46:09 +0100


base/vistat/vistat.c
/****************************************************************
 *
 * Copyright (C) Max Planck Institute 
 * for Human Cognitive and Brain Sciences, Leipzig
 *
 * Author Gabriele Lohmann, 2004, <lipsia@cbs.mpg.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * $Id: vistat.c 3629 2009-08-20 17:04:30Z proeger $
 *
 *****************************************************************/

/* From the V library: */
#include "viaio/Vlib.h"
#include "viaio/mu.h"
#include "viaio/option.h"
#include "viaio/os.h"
#include "viaio/VImage.h"

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#define ABS(x) ((x) > 0 ? (x) : -(x))


VImage
VImStats(VImage src)
{
  int b,r,c,n,i;
  double v,sum1,sum2,nx,xmin,xmax,tiny=1.0e-6;
  double mean,sigma;

  xmin = VRepnMaxValue(VDoubleRepn);
  xmax = VRepnMinValue(VDoubleRepn);

  sum1 = sum2 = nx = 0;
  for (b=0; b<VImageNBands(src); b++) {
    for (r=0; r<VImageNRows(src); r++) {
      for (c=0; c<VImageNColumns(src); c++) {
	v = VGetPixel(src,b,r,c);
	if (v < xmin) xmin = v;
	if (v > xmax) xmax = v;
	if (ABS(v) < tiny) continue;
	sum1 += v;
	sum2 += v*v;
	nx++;
      }
    }
  }
  if (nx < 1) VError(" no non-zero voxels found");
  mean = sum1/nx;
  sigma = sqrt((sum2 - nx * mean * mean) / (nx - 1.0));

  
  fprintf(stderr,"\n image stats:\n");
  fprintf(stderr," min = %g, max= %g\n",xmin,xmax);
  fprintf(stderr," mean= %g, std= %g\n\n",mean,sigma);
}




int main (int argc, char **argv)
{
  FILE *in_file;
  VAttrList list;
  VAttrListPosn posn;
  VImage image;
  char prg[50];	
  sprintf(prg,"vistat V%s", getVersion());
  fprintf (stderr, "%s\n", prg);

  /* Parse command line arguments and identify the input and output files: */
  VParseFilterCmd (NULL, 0, argc, argv, & in_file,NULL);

  /* Read the input file: */
  if (! (list = VReadFile (in_file, NULL)))
    exit (EXIT_FAILURE);

  /* For each image in the input file: */
  for (VFirstAttr (list, & posn); VAttrExists (& posn); VNextAttr (& posn)) {
    if (VGetAttrRepn (& posn) != VImageRepn) continue;
    VGetAttrValue (& posn, NULL, VImageRepn, & image);
    VImStats(image);
  }
}
